Rust Foundation: April 2024 Update (#6)
JoelMarcey/alpha-omega: alpha/engagements/2024/Rust Foundation/update-2024-04.md

April was an odd month with a bunch of travel, vacation and sickness across the Rust Foundation technology team. That said, there were still some security-related updates to share.

## Engineering

> ### Standard Library Security Advisory
> 
> The Rust Security Response WG was notified that the Rust standard library did not properly escape arguments when invoking batch files (with the `bat` and `cmd` extensions) on Windows using the `Command` API. An attacker able to control the arguments passed to the spawned process could execute arbitrary shell commands by bypassing the escaping.
> 
> Due to the complexity of `cmd.exe`, the Security Response WG didn't identify a solution that would correctly escape arguments in all cases. To maintain API guarantees, they improved the robustness of the escaping code, and changed the Command API to return an InvalidInput error when it cannot safely escape an argument. This error will be emitted when spawning the process.
> 
> A [blog post](https://blog.rust-lang.org/2024/04/09/cve-2024-24576.html) detailing the entire process was written.

Reviewer: ```suggestion
### Standard Library Security Advisory

The Rust Security Response WG was notified that the Rust standard library (among many other software development ecosystems) did not properly escape arguments when invoking batch files (with the `bat` and `cmd` extensions) on Windows using the `Command` API. An attacker able to control the arguments passed to the spawned process could execute arbitrary shell commands by bypassing the escaping.

Due to the complexity of `cmd.exe`, the Security Response WG didn't identify a solution that would correctly escape arguments in all cases. To maintain API guarantees, they improved the robustness of the escaping code, and changed the Command API to return an `InvalidInput` error when it cannot safely escape an argument. This error will be emitted when spawning the process.

A [blog post](https://blog.rust-lang.org/2024/04/09/cve-2024-24576.html) detailing the entire process was written.
```

### R2D2 Replaced by Deadpool

> The `r2d2` database connection pools has been totally [removed](https://github.com/rust-lang/crates.io/pull/8442) from the crates.io backend code. All of the code has been migrated to `deadpool` async pools. This is important because `r2d2` is obsolete and unmaintained and `deadpool` is active and supports `diesel-async`. 

Reviewer: ```suggestion
### crates.io: Unmaintained `r2d2` dependency replaced by `deadpool`

The `r2d2` database connection pool library has been [removed](https://github.com/rust-lang/crates.io/pull/8442) from the crates.io backend code, since it appears to be unmaintained. It has been replaced by an async database connection pool library called `deadpool`, which is actively maintained and supported by `diesel-async`.

> ### crates.io Test Suite Now Uses Async/Await
> 
> crates.io now uses [async testing](https://github.com/rust-lang/crates.io/pull/8498) via Tokio. 

Reviewer: ```suggestion
### crates.io: Test suite now uses async/await

The crates.io test suite was [migrated to async tests](https://github.com/rust-lang/crates.io/pull/8498) to make it easier to use async-only libraries in the test suite.
```

> 

Reviewer: ```suggestion

### crates.io: Database performance improvements

Now that the download traffic is no longer hitting the crates.io API servers, the crates.io team began to investigate performance issues and their potential solutions for the remaining API calls. The reverse dependencies endpoint in particular is not scaling well with the increased number of packages on crates.io. A [`default_versions` database table](https://github.com/rust-lang/crates.io/pull/8484) was introduced as a first step to address some of these performance concerns.
```
